💾 Created and updated the models for moderation

sleeping-dev opened the pull request:
Took care of the backend changes for #21:

- I removed the IsAdmin from Users and added the enum 'UserType' (felt a bit more fitting than UserStatus)
- UserType has User(=1), Moderator(=2) and Admin(= 3) as values, in the database they're stored as ints.
- When creating a new user, the UserType is set to User by default.
- During migration, the UserType column will be filled with 1 when created, so all testing accounts that already exist will be users.
- Added Reports, Infractions, and made the other changes requested by #21.

@Breadkenty, in #21 you mentioned a property `type[required]` for the `Infraction` class, could you elaborate on what that is for? Or is that a leftover of the design process? I left it out for the time being, since I wasn't sure what it was.

skittlz444 commented on Smash_Combos.Domain/Models/Infraction.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public int ModeratorId { get; set; }
+
+        public int BanDuration { get; set; }
```
This should be nullable incase the infraction is a warning.

skittlz444 commented on Smash_Combos.Domain/Models/Infraction.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public int BanDuration { get; set; }
+
+        public int Points { get; private set; }
```
Should this also be nullable?

Breadkenty replied:
I think when a user gets infracted, they will always get a point, including bans. What do you think?

sleeping-dev replied:
If we have points in bans too then we should keep it that way. But do we need them? Wouldn't a mod see the user's accumulated points and then decide to ban them? Then I don't think we would need to add additional points on the bans themselves.
I'm leaning towards nullable, personally. Good thinking.

Breadkenty replied:
What if the user gets banned, temporarily. Say 2 weeks. Would they come back with no points for whatever the offense is? or would it make more sense for the moderator to infract AND THEN ban the user?

sleeping-dev replied:
The infraction would still exist, so they would still have the points to their name, even after the ban. The way I thought about it was that the moderator does their work putting out infractions, until they notice that a user has a high amount of infractions/points, and then they issue a ban, setting the length to whatever they feel is long enough. When a user comes back after a ban and gets reported AGAIN, the moderator would see their high point count and would likely ban them again, longer/permanent this time. The bans themselves would not have points, they're just there to _punish_ a user that racks up a lot of points. Was that what you were envisioning?

Breadkenty replied:
My initial thought was that bans will also give points as a warning would. But honestly, your approach feels much better. I was just trying to reduce the number of steps for UI purposes, but again, this approach feels better. Lets make it nullable.

sleeping-dev replied:
Alright, I'll change it real quick.

skittlz444 commented on Smash_Combos.Persistence/Migrations/20200816160856_UpdatedForModeration.cs:
```diff
@@ -0,0 +1,112 @@
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.DropColumn(
```
You should probably do this after creating the new column, and migrate over any existing admins so that they're still admins in the new column

skittlz444 commented on Smash_Combos.Persistence/Migrations/20200816160856_UpdatedForModeration.cs:
```diff
@@ -0,0 +1,112 @@
+                        .Annotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.IdentityByDefaultColumn),
+                    UserId = table.Column<int>(nullable: false),
+                    ModeratorId = table.Column<int>(nullable: false),
+                    BanDuration = table.Column<int>(nullable: false),
```
Should be nullable?

skittlz444 commented on Smash_Combos.Persistence/Migrations/20200816160856_UpdatedForModeration.cs:
```diff
@@ -0,0 +1,112 @@
+                {
+                    table.PrimaryKey("PK_Infractions", x => x.Id);
+                    table.ForeignKey(
+                        name: "FK_Infractions_Users_UserId",
```
There also seems to be missing a constraint for moderator id

sleeping-dev replied:
I'm so used to migrations taking care of everything that I missed this and the other foreignkey missing. 😅 

I'll set the ReferentialAction for this constraint to SetNull, since we don't want bans to disappear if a moderator deletes their account.

skittlz444 replied:
@sleeping-dev If you set this to SetNull, you need to make ModeratorId nullable.

skittlz444 commented on Smash_Combos.Persistence/Migrations/20200816160856_UpdatedForModeration.cs:
```diff
@@ -0,0 +1,112 @@
+                {
+                    Id = table.Column<int>(nullable: false)
+                        .Annotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.IdentityByDefaultColumn),
+                    UserId = table.Column<int>(nullable: false),
```
There seems to be no foreign keys for these user columns

skittlz444 commented on Smash_Combos.Persistence/Migrations/20200816160856_UpdatedForModeration.cs:
```diff
@@ -0,0 +1,112 @@
+            migrationBuilder.DropTable(
+                name: "Reports");
+
+            migrationBuilder.DropColumn(
```
Similar thing to the changes required to migrate up, you should create the column first and migrate admins back

sleeping-dev commented:
On my way to work now, will update when I get home.

Breadkenty commented:
@sleeping-dev A lot of this looks great, @skittlz444 has some great feedback there. As for the `type` sorry I didn't make that part clear! Also, it's not on the `Report` model, but instead for the `Infraction`. I wanted a way for mods to categorize what type of infraction they are giving along with the comments. So for example:

### Infraction
`Type`: Spamming
`Body`: User was spamming emojis in the comment section
`Points`: 1

Hope this is clear, sorry I didn't get back sooner I got caught up with some stuff.

sleeping-dev commented:
Had a long day, sorry for the slow progress. I sat down just now when I got home and came across a problem during testing:
Everything works fine, so I wanted to see if the cascading delete's worked correctly by submitting a combo and writing a report into the database. But for some reason, when there's a combo in the database, the GetCharacters-Api call messes up, I get a JsonException:
`A possible object cycle was detected which is not supported. This can either be due to a cycle or if the object depth is larger than the maximum allowed depth of 32.`

I tracked this down to be because Combos have a `User` object with them, and since Users have a List of Combos now this is an endless loop and Json doesn't like that 😄 
We already have the UserId in the Combo object, so I don't think we need the User too. I went ahead and removed them, but the frontend depends on the User object being sent with a combo.

@Breadkenty I think we should refactor the front end so instead of expecting a User object, it expects the UserId and then calls the Api to get the username of the combo creator.

The same circular reference exists in Comments (Users have a List of Comments, Comments have a User reference), I'm a bit surprised we haven't encountered this before? Maybe it's a bit different and I'm missing something, but I don't think we need the User reference in the Comment either, we can just get the Username, etc. via the UserId that the comment has.

sleeping-dev commented:
Alright, added the necessary changes, let me know if you find anything that I missed! @Breadkenty I changed Character.jsx and CharacterCombo.jsx to not show the username of the authors of combos/comments, since up until now they used the userobject that came with the combo/comment for that. Do you think we could tweak this so those pages use the UserId that combos/comments have to make an Api call to the UsersController and get the username that way?

skittlz444 commented:
@sleeping-dev Another thing I've just notice (should've picked it up in the first review) is that you aren't adding navigation properties to the domain (e.g. `public User Moderator`). These should be used over ids generally in domain driven design, the only Id usually included in a domain model is the id of the model itself (e.g. ReportId on the Report model). All other things should be using navigation properties.

skittlz444 commented:
@sleeping-dev @Breadkenty a word on that infinite json loop. This is caused because the api is returning domain models as json. It makes sense for these models to contain the properties, so removing them doesn't make sense. The best approach is for those responses causing loops, convert those to core use cases where you have a better control over their esponse type, and can explicitly say that the "GetCombo" will return the shallow user (no combos/comments or any other data that combo doesn't need) and that "GetUser" will return the shallow combo (no user).

sleeping-dev commented:
I actually thought having Ids wpuld be the cleaner approach, but you're making sense, I'll rework it when I'm off work.

sleeping-dev commented:
OK! Sorry for taking so long everyone, but this should take care of adding/updating the backend models. I modified the API Controllers where necessary to avoid the endless JSON loop problem (thanks to @skittlz444 again for helping out), so this is up and running with the new/changed models.

We should move the other API logic into the Core project as well, but I figured I shouldn't overload this pull request with all the API changes since @Breadkenty is waiting with the frontend stuff already.

@skittlz444 I hope what I did in the Core was OK, I figured since GetCharacter and GetCharacters return the same base-object (just one of them a list), I use the same Profile/Response object, since they pretty much hold the same data. If this is breaking the pattern/is against the idea of MediatR let me know.

skittlz444 commented:
@sleeping-dev Will review in a bit, but on your comment about cutting over the rest of the controllers. This can happen in stages, we don't need to do it immediately. However, since we'll probably have some time this week whist @Breadkenty is away, it's probably a good time to cut over as much as we can. I'd say wait to do it in a new pull request and issue though.

skittlz444 commented:
@sleeping-dev Generally you shouldn't share response objects, but for now it's ok. We can always modify that when needed later.

skittlz444 commented on Api/Controllers/CharactersController.cs:
```diff
@@ -20,30 +24,33 @@ public class CharactersController : ControllerBase
+            if (response == null)
             {
-                return await _context.Characters.Where(character => character.Name.ToLower().Contains(filter) || character.VariableName.ToLower().Contains(filter)).Include(character => character.Combos).ThenInclude(combo => combo.User).ToListAsync();
+                // Return a `404` response to the client indicating we could not find a combo with this id
```
Copy paste error

skittlz444 commented on Api/Controllers/CharactersController.cs:
```diff
@@ -20,30 +24,33 @@ public class CharactersController : ControllerBase
+                return NotFound();
             }
+
+            //  Return the combo as a JSON object.
```
Copy paste error

skittlz444 commented on Smash_Combos.Core/Cqrs/Characters/CharacterResponse.cs:
```diff
@@ -0,0 +1,17 @@
+        public string VariableName { get; set; }
+        public int YPosition { get; set; }
+        public decimal ReleaseOrder { get; set; }
+        public List<Combo> Combos { get; set; }
```
This actually made me realise I also made this mistake (copy paste whoops). We shouldn't use the domain entity as part of the response object, this will probably still cause the json loop error. You should create another class of "ComboDto" that contains just the information needed from the Combo entity, for the character response. i.e. if only 3 of the properties are used, then only get those 3.
You'll also have to create another map in the profile to make it work.

sleeping-dev replied:
If I understand correctly that means we'll have different ComboDtos for different requests, right?

skittlz444 replied:
Yup that's exactly right

skittlz444 commented on Smash_Combos.Core/Cqrs/Characters/GetCharacter/GetCharacterRequestHandler.cs:
```diff
@@ -0,0 +1,58 @@
+                if (character == null)
+                    return null;
+
+                if(character.Combos?.Count > 0)
```
If you update the response to not use the domain entity, then this won't be needed.

skittlz444 commented on Smash_Combos.Core/Cqrs/Combos/GetCombo/GetComboRequestHandler.cs:
```diff
@@ -22,11 +22,24 @@ public GetComboRequestHandler(IDbContext context, IMapper mapper)
             if (combo == null)
                 return null;
 
+            if(combo.Comments?.Count > 0)
```
If you do the same thing with this response object, as the other, then this also gets removed.

skittlz444 commented on Smash_Combos.Domain/Models/Infraction.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Smash_Combos.Domain.Models
+{
+    public class Infraction
```
Infraction should have `public User User` and `public User Moderator` rather than the ids

skittlz444 commented on Smash_Combos.Domain/Models/Report.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Smash_Combos.Domain.Models
+{
+    public class Report
```
Report should have `public User User` and `public User Reporter` rather than the ids

skittlz444 reviewed (changes requested):
Just a few things with the Core response, which I also made the mistake of doing the same thing. Also a couple of things with the domain entities.

sleeping-dev commented:
Thanks for the review, I made the changes you suggested. I just added the Dtos that I needed to their respective Method Folders for clarity, let me know if I should change it.

I also split up the response object again, while implementing this I saw how there's no real sense in trying to use one response object for multiple requests, since the properties we actually need change rapidly from request to request.

skittlz444 reviewed (approved):
Looks good! Good to go imo @Breadkenty.

Breadkenty commented:
Awesome job y'all! @sleeping-dev @skittlz444
